app/ide-desktop/lib/dashboard/src/authentication/src/authentication/providers/auth.tsx

```diff
@@ -0,0 +1,238 @@
+    accessToken: string;
+    /** User's email address. */
+    email: string;
+    /** User's organization information. */
+    organization: backendService.Organization;
+}
+
+/** Object containing the currently signed-in user's session data, if the user has not yet set their
+ * username.
+ *
+ * If a user has not yet set their username, they do not yet have an organization associated with
+ * their account. Otherwise, this type is identical to the `Session` type. This type should ONLY be
+ * used by the `SetUsername` component. */
+export interface PartialUserSession {
+    /** A type-hint for TypeScript to be able to disambiguate between this interface and other
+     * `UserSession` variants. */
+    state: "partial";
+    /** User's JSON Web Token (JWT), used for authenticating and authorizing requests to the API. */
+    accessToken: string;
+    /** User's email address. */
+    email: string;
+}
+
+
+
+// ===================
+// === AuthContext ===
+// ===================
+
+/** Interface returned by the `useAuth` hook.
+ *
+ * Contains the currently authenticated user's session data, as well as methods for signing in,
+ * signing out, etc. All interactions with the authentication API should be done through this
+ * interface.
+ *
+ * See {@link Cognito} for details on each of the authentication functions. */
+interface AuthContextType {
+    /** Session containing the currently authenticated user's authentication information.
+     *
+     * If the user has not signed in, the session will be `undefined`. */
+    session: UserSession | undefined;
+}
+
+/** Create a global instance of the `AuthContextType`, that will be re-used between all React
+ * components that use the `useAuth` hook.
+ *
+ * # Safety of Context Initialization
+ *
+ * An `as ...` cast is unsafe. We use this cast when creating the context. So it appears that the
+ * `AuthContextType` can be unsafely (i.e., only partially) initialized as a result of this.
+ *
+ * So it appears that we should remove the cast and initialize the context as `undefined` instead.
+ *
+ * **However**, initializing a context the existing way is the recommended way to initialize a
+ * context in React.  It is safe, for non-obvious reasons. It is safe because the `AuthContext` is
+ * only accessible through the `useAuth` hook.
+ *
+ * 1. If the `useAuth` hook is called in a component that is a child of an `AuthProvider`, then the
+ * context is guaranteed to be initialized, because the `AuthProvider` constructor is what
+ * initializes it. So the cast is safe.
+ * 2. If the `useAuth` hook is called in a component that is not a child of an `AuthProvider`, then
+ * the hook will throw an error regardless, because React does not support using hooks outside of
+ * their supporting providers.
+ *
+ * So changing the cast would provide no safety guarantees, and would require us to introduce null
+ * checks everywhere we use the context. */
```

Comment:
Oh well. This comment explains a lot, but Im sad you spent so much time writing it. It would be enough to write that this is the recommended way of doing it by React with a link to it, but huh, Im sad you were facing this issue.